engine/runtime/src/test/java/org/enso/interpreter/test/PrintTest.java

```diff
@@ -0,0 +1,244 @@
+        IO.println a
+    """;
+
+    checkPrint(code, "(Error: (My_Error.Error 1))");
+  }
+
+  @Test
+  public void testPrintToTextHasWarnings() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+    from Standard.Base.Warning import Warning
+
+    test =
+        a = Warning.attach "Warning" "FOOBAR"
+        IO.println a
+    """;
+
+    checkPrint(code, "FOOBAR");
+  }
+
+  @Test
+  public void testPrintToTextHasWarnings2() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+    from Standard.Base.Warning import Warning
+
+    test =
+        a = Warning.attach "Warning" 42
+        IO.println a
+    """;
+
+    checkPrint(code, "42");
+  }
+
+  @Test
+  public void testPrintToTextHasWarnings3() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+    from Standard.Base.Warning import Warning
+
+    type My_Object
+        Value x
+        
+        to_text self = "MyObj{" + self.x.to_text + "}"
+
+    test =
+        a = Warning.attach "Warning" (My_Object.Value 42)
+        IO.println a
+    """;
+
+    checkPrint(code, "MyObj{42}");
+  }
+
+  @Test
+  public void testPrintToTextHasWarnings4() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+    from Standard.Base.Warning import Warning
+
+    type My_Object
+        Value x
+        
+        to_text self =
+            res = "MyObj{" + self.x.to_text + "}"
+            Warning.attach "Warning2" res
+
+    test =
+        a = Warning.attach "Warning" (My_Object.Value 42)
+        IO.println a
+    """;
+
+    checkPrint(code, "MyObj{42}");
+  }
+
+  @Test
+  public void testPrintToTextTypeError() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+
+    type My_Object
+        Value x
+        
+        to_text self = 100
+
+    test =
+        a = My_Object.Value 42
+        IO.println a
+    """;
+
+    checkPrint(code, "100");
+  }
+
+  @Test
+  public void testPrintToTextTypeErrorAndWarnings() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+    from Standard.Base.Warning import Warning
+
+    type My_Object
+        Value x
+        
+        to_text self =
+            res = 100
+            Warning.attach "Warning2" res
+
+    test =
+        a = Warning.attach "Warning2" (My_Object.Value 42)
+        IO.println a
+    """;
+
+    checkPrint(code, "100");
+  }
+
+  // This test documents the current, but undesirable behaviour. It should be updated when it is fixed.
+  @Test
+  public void testPrintToTextStaticMethod() throws Exception {
+    final String code = """
+    import Standard.Base.IO
+
+    type My_Object
+        Value x
+        
+        to_text self = "MyObj{" + self.x.to_text + "}"
+
+    test =
+        a = My_Object
+        IO.println a
+    """;
+
+    checkPrint(code, "My_Object.type.to_text[test:6-16]");
+    // We would want the following result:
+    //checkPrint(code, "My_Object");
```

Comment:
This is now being discussed here: https://github.com/orgs/enso-org/discussions/6216

Reply:
Still, better to get `My_Object.type.to_text[test:6-16]` than a Type Error 😁

Reply:
How can I add the `@Specialization` for `Type` so that it doesn't clash with:
```
  @Specialization(guards = "!strings.isString(message)")
  Object doPrint(
      VirtualFrame frame,
      State state,
      Object message,
```
?

`Type <: Object` so I need to somehow add to this specialization a condition that `message </: Type`, how do I do this in Truffle? Just `guards = "!(message instanceof Type)"`?

Reply:
I would add:
```java
@Specialization
Object doPrintTypeName(VirtualFrame frame, State state, Type message) {
...
}
```
and possibly expanded the `guards = { ...., "!(message instanceof Type)" }` as you propose

Reply:
Ok, I will try that

Reply:
After all it turned out to be a bit more complicated, and I think there's little sense in adding such a specific (and tbh not _that_ often used) specialization which just fixes a small part of the problem (it would fix `IO.println Type` but if I do `IO.println "Hello: "+Type.to_text` it would still be broken and I think the latter is used even more often).

The whole problem will be fixed anyway by #6300 which is scheduled to be done soon, so there is no sense in keeping a specialized fix for a small case (especially as I'd need a bit more time to get it right).